pyne/ace.pyx

```diff
@@ -1573,8 +1711,324 @@ class Reaction(object):
+        """
+        # 1. sample energy
+        edist = self.energy_dist
+        if edist.law == 3:
+            # Inelastic level scattering
+            A = self.table.awr 
+            Q = self.Q
+            E_out = (A / (A + 1.0)) ** 2 * (e - (A + 1.0) / A * Q)
+            mu = self.sample_mu(e)
+            return[mu, E_out]
+            
+        elif edist.law == 4:
+            #Continuous Tabular Distribution 
+            if hasattr(edist, 'INT'):
+                assert len(edist.INT) == 1, 'Multiple interpolation regions not yet supported ' \
+                                            'for continuous tabular energy distributions.'
+                histogram_interp = (edist.INT[0] == 1)
+            else:
+                histogram_interp = False
+                
+            # Find energy bin and calculate interpolation factor -- if the energy is
+            # outside the range of the tabulated energies, choose the first or last bins 
+            if e <= edist.energy_in[0]:
+                i = 0
+                f = 0.0  
+            elif e >= edist.energy_in[-1]:
+                i = len(edist.energy_in) -2 
+                f = 1.0  
+            else:
+                i = np.searchsorted(edist.energy_in, e) - 1
+                f = (e - edist.energy_in[i]) / (edist.energy_in[i+1] - edist.energy_in[i])
+            
+            # Sample between the ith and (i+1)th bin
+            if (histogram_interp):
+                if e >= edist.energy_in[-1]:
+                    l = i + 1  
+                else:
+                    l = i 
+            else:
+                if (f > rand()):
+                    l = i + 1 
+                else:
+                    l = i 
+                    
+            # check for discrete lines present
+            assert edist.nd[l] == 0, 'Discrete lines in continuous tabular ' \
+                                     'distribution not yet supported.'
+                                     
+            # Interpolation for energy E1 and EK
+            E_i_1 = edist.energy_out[i][0]
+            E_i_K = edist.energy_out[i][-1]
+            
+            E_i1_1 = edist.energy_out[i+1][0]
+            E_i1_K = edist.energy_out[i+1][-1]
+            
+            E_1 = E_i_1 + f*(E_i1_1 - E_i_1)
+            E_K = E_i_K + f*(E_i1_K - E_i_K)
+            
+            E_out = tabular_sample(edist.intt[l], edist.energy_out[l], \
+                                   edist.pdf[l], edist.cdf[l])
+            
+            # Interpolate between incident energy bins i and i + 1
+            if not histogram_interp:
+                if (l == i):
+                    E_out = E_1 + (E_out - E_i_1)*(E_K - E_1)/(E_i_K - E_i_1)
+                else:
+                    E_out = E_1 + (E_out - E_i1_1)*(E_K - E_1)/(E_i1_K - E_i1_1)
+                    
+            # Sample mu 
+            mu = self.sample_mu(e)
+            return [mu, E_out]
+                    
+        elif edist.law == 44:
```

Comment:
This function is getting pretty unweildy. Please break thi sup into sub-functions.